tests/neo4j/txrun.py

```diff
@@ -118,18 +122,143 @@ def test_should_fail_run_in_a_rollbacked_tx(self):
+        if self.driver_supports_features(types.Feature.TMP_RESULT_KEYS):
+            self.assertEqual(result.keys(), ["x"])
+        values = []
+        for _ in range(2):
+            record = result.next()
+            self.assertIsInstance(record, types.Record)
+            values.append(record.values[0])
+        tx.commit()
+
+        # TODO: what should happen here?
+        # options:
+        #   - Don't fetch any further records but return what is buffered rn
+        #   - throw exception
+        #   - Commit should've buffered all records and return them now
+
+    def test_parallel_queries(self):
+        def _test():
+            self._session = self._driver.session("w", fetchSize=2)
+            tx = self._session.beginTransaction()
+            result1 = tx.run("UNWIND [1,2,3,4] AS x RETURN x")
+            result2 = tx.run("UNWIND [5,6,7,8] AS x RETURN x")
+            if self.driver_supports_features(types.Feature.TMP_RESULT_KEYS):
+                self.assertEqual(result1.keys(), ["x"])
+                self.assertEqual(result2.keys(), ["x"])
+            if invert_fetching:
+                values2 = list(map(lambda rec: rec.values[0], result2))
+                values1 = list(map(lambda rec: rec.values[0], result1))
+            else:
+                values1 = list(map(lambda rec: rec.values[0], result1))
+                values2 = list(map(lambda rec: rec.values[0], result2))
+            tx.commit()
+            self.assertEqual(values1, list(map(types.CypherInt, (1, 2, 3, 4))))
+            self.assertEqual(values2, list(map(types.CypherInt, (5, 6, 7, 8))))
+            self._session.close()
+            self._session = None
+
+        for invert_fetching in (True, False):
+            with self.subTest("inverted" if invert_fetching else "in_order"):
+                _test()
+
+    def test_interwoven_queries(self):
+        def _test():
+            self._session = self._driver.session("w", fetchSize=2)
+            tx = self._session.beginTransaction()
+            result1 = tx.run("UNWIND [1,2,3,4] AS x RETURN x")
+
+            if run_q2_before_q1_fetch:
+                result2 = tx.run("UNWIND [5,6,7,8] AS y RETURN y")
+
+            if self.driver_supports_features(types.Feature.TMP_RESULT_KEYS):
+                self.assertEqual(result1.keys(), ["x"])
+            values1 = [result1.next().values[0]]
+
+            if not run_q2_before_q1_fetch:
+                result2 = tx.run("UNWIND [5,6,7,8] AS y RETURN y")
+            if self.driver_supports_features(types.Feature.TMP_RESULT_KEYS):
+                self.assertEqual(result2.keys(), ["y"])
+            values2 = list(map(lambda rec: rec.values[0], result2))
+
+            self.assertEqual(values2, list(map(types.CypherInt, (5, 6, 7, 8))))
+            tx.commit()
+            self.assertIsInstance(result2.next(), types.NullRecord)
+            self.assertEqual(values1, [types.CypherInt(1)])
+            # TODO: what should result1.next() result in?
```

Comment:
Or It should return all the already buffered records until and then fails when trying fetching more results or it should fail immediately.

Reply:
True. Keep that in mind for the next unification meeting, please!